File: src/main/java/com/ggardet/core/enumeration/ValueOfEnum.java
package com.ggardet.core.enumeration;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;
import org.springframework.modulith.NamedInterface;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.*;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

@Target({METHOD, FIELD, ANNOTATION_TYPE, CONSTRUCTOR, PARAMETER, TYPE_USE})
@Retention(RUNTIME)
@Documented
@Constraint(validatedBy = ValueOfEnumValidator.class)
@NamedInterface("core-value-of-enum-spi")
public @interface ValueOfEnum {
    Class<? extends Enum<?>> enumClass();

    String regexp();

    String field();

    String message() default "The field {field} must equals one of these values : {regexp}";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};
}

File: src/main/kotlin/com/ggardet/core/enumeration/ValueOfEnum.kt
package com.ggardet.core.enumeration

import jakarta.validation.Constraint
import jakarta.validation.Payload
import org.springframework.modulith.NamedInterface
import kotlin.annotation.AnnotationRetention.RUNTIME
import kotlin.annotation.AnnotationTarget.*
import kotlin.reflect.KClass

@Target(FUNCTION, PROPERTY_GETTER, PROPERTY_SETTER, FIELD, ANNOTATION_CLASS, CONSTRUCTOR, VALUE_PARAMETER, TYPE)
@Retention(RUNTIME)
@MustBeDocumented
@Constraint(validatedBy = [ValueOfEnumValidator::class])
@NamedInterface("core-value-of-enum-spi")
annotation class ValueOfEnum(
    val enumClass: KClass<out Enum<*>>,
    val regexp: String,
    val field: String,
    val message: String = "The field {field} must equals one of these values : {regexp}",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = []
)